include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -157,32 +161,291 @@ class NBL_API2 IShaderCompiler : public core::IReferenceCounted
+					std::string filepathHint;
+					uint64_t codeByteSize = 0;
+					uint64_t offset = 0; // Offset into the serialized .bin for the Cache where code starts
+				};
+
+				std::string mainFileContents;
+				SCompilerData compilerData;
+				// Keeping this one commented out. Could be useful for lazy-loading of dependencies if it takes up too much memory
+				// uint64_t entryID;
+
+				dependency_container_t dependencies;
+				mutable CPUShaderCreationParams shaderParams;
+				bool serialized = false;
+				mutable core::smart_refctd_ptr<asset::ICPUShader> value = nullptr;
+			};
+
+			inline void insert(SEntry&& entry)
+			{
+				m_container.insert(std::move(entry));
 			}
 
-			virtual IShader::E_CONTENT_TYPE getCodeContentType() const { return IShader::E_CONTENT_TYPE::ECT_UNKNOWN; };
+			// can move to .cpp and have it not inline
+			inline core::smart_refctd_ptr<asset::ICPUShader> find(const SEntry& mainFile, CIncludeFinder* finder)
+			{
+				auto foundRange = m_container.equal_range(mainFile);
+				for (auto& found = foundRange.first; found != foundRange.second; found++)
+				{
+					bool allDependenciesMatch = true;
+					// go through all dependencies
+					for (auto i = 1; i != found->dependencies.size(); i++)
+					{
+						const auto& dependency = found->dependencies[i];
+
+						IIncludeLoader::found_t header;
+						if (dependency.standardInclude)
+							header = finder->getIncludeStandard(dependency.requestingSourceDir, dependency.identifier);
+						else
+							header = finder->getIncludeRelative(dependency.requestingSourceDir, dependency.identifier);
+
+						if (header.hash != dependency.hash || header.contents != dependency.contents)
+						{
+							allDependenciesMatch = false;
+							break;
+						}
+					}
+					if (allDependenciesMatch) {
+						if (!found->value) { // Load shader if not already loaded
+							// If the Cache has no storageBuffer, then it's fresh new (not picked up from serialization): all compiled shaders should be loaded
+							// If codeByteSize is 0, then the Entry found lives in the Cache but not in the serialization, so the Entry was generated in the current runtime:
+							// This means the shader should be loaded
+							assert(storageBuffer && found->shaderParams.codeByteSize != 0);
+							auto code = core::make_smart_refctd_ptr<ICPUBuffer>(found->shaderParams.codeByteSize);
+							memcpy(code->getPointer(), (uint8_t*)storageBuffer->getPointer() + CONTAINER_JSON_SIZE_BYTES + containerJsonSize + found->shaderParams.offset, found->shaderParams.codeByteSize);
+							found->value = core::make_smart_refctd_ptr<ICPUShader>(std::move(code), found->shaderParams.stage, found->shaderParams.contentType, std::move(std::string(found->shaderParams.filepathHint)));
+						}
+						return found->value;
+					}	
+				}
+				return nullptr;
+			}
+
+			CCache() {}
+
+			// TODO: add methods as needed, e.g. to serialize and deserialize to/from a pointer
+			std::vector<uint8_t> serialize();
```

Comment:
can you somehow work out an upper bound on the serialization size?

If so you could just have a "serialize to a pointer" method

If you can't then return an `ICPUBuffer` with an allocator that hijacks a `core::vector<uint8_t>` or `core::vector<uint64_t>`

Reply:
I couldn't work out an upper bound, but I'm not getting what was meant by the latter.

Reply:
So I'll extend ICPUBuffer with some custom CVectorBuffer or something similar, that holds its own vector. What methods do I need to override in this case?

The same question goes for the CFileBackedCPUBuffer

Reply:
Only the methods which are virtual can be overriden so
```cpp
template<typename T, typename Allocator>
class CVectorCPUBuffer final : public ICPUBuffer
{
   public:
       inline CVectorCPUBuffer(core::vector<T,Allocator>&& _vec) : m_vec(std::move(_vec))
       {
           data = m_vec.data();
       }
   
   protected:
       virtual inline ~CVectorCPUBuffer()
       {
           freeData();
       }
       inline void freeData() override
       {
           m_vec.clear();
           ICPUBuffer::data = nullptr;
       }
   
       core::vector<T,Allocator> m_vec;
};
```